### PR TITLE
feat(abstract-things): new type definition

### PR DESCRIPTION
**Pull request by Shaquu**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [x] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## abstract-things ([<kbd>unpkg</kbd>](https://unpkg.com/browse/abstract-things@latest/))
was missing the following properties:
1. Discovery
1. Polling
1. State
1. RestorableState
1. ErrorState

<details>
<summary>as well as these 12 other properties...</summary>
<p>Storage, Children, Nameable, EasyNameable, Power, SwitchablePower, Mode, SwitchableMode, BatteryLevel, ChargingState, AutonomousCharging, Placeholder</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1f2c6d91f89171314c33e08cce10aab695f2777b
</p>

File: types/abstract-things/thing.d.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import EventEmitter from './events';
+
+export interface EmitEventOptions {
+    multiple: boolean;
+}
+
+export class Thing {
+    constructor(...args: any);
+    new(...input: any[]): any;
```
Comment (peterblazejewicz) [inline types/abstract-things/thing.d.ts:9]:
```suggestion
```
why this additional constructor?

File: types/abstract-things/thing.d.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    get id(): string | null;
+    set id(identifier: string | null);
+
+    init(): Promise<undefined | this>;
+    initCallback(): Promise<void>;
+
+    destroy(): Promise<void>;
+    destroyCallback(): Promise<void>;
+
+    emitEvent(event: string, data: unknown, options?: EmitEventOptions): void;
```
Comment (peterblazejewicz) [inline types/abstract-things/thing.d.ts:20]:
```suggestion
    emitEvent(event: string, data?: unknown, options?: EmitEventOptions): void;
```
the param is optional IMO:
https://github.com/thingbound/abstract-things/blob/440404b3b7f72047ad154f49f1f289eeda736232/thing.js#L72

File: types/abstract-things/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Type definitions for abstract-things 0.9
+// Project: https://github.com/thingbound/abstract-things
+// Definitions by: Tadeusz Wyrzykowski <https://github.com/Shaquu>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
```
Comment (peterblazejewicz) [inline types/abstract-things/index.d.ts:6]:
```suggestion
```
you're not using any of NodeJS types directly, why this import?

File: types/abstract-things/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+import { Thing } from './thing';
+
+export { Thing };
```
Comment (peterblazejewicz) [inline types/abstract-things/index.d.ts:10]:
```suggestion
import Thing = require('./thing');

export { Thing };
```

File: types/abstract-things/thing.d.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import EventEmitter from './events';
+
+export interface EmitEventOptions {
+    multiple: boolean;
+}
+
+export class Thing {
```
Comment (peterblazejewicz) [inline types/abstract-things/thing.d.ts:7]:
```suggestion
declare namespace Thing {
    interface EmitEventOptions {
        multiple: boolean;
    }
}

declare class Thing {
```

File: types/abstract-things/thing.d.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    debug(): void;
+    matches(...tags: string[]): boolean;
+
+    static type<T>(func: T): T & Thing;
+    static mixin<T>(func: T): T & Thing;
+    static mixinDynamic<T>(...mixins: T[]): T & Thing;
+
+    extendWith<T>(...mixins: T[]): T & Thing;
+}
```
Comment (peterblazejewicz) [inline types/abstract-things/thing.d.ts:34]:
```suggestion
}

export = Thing;
```

**Comment by Shaquu** [comment]
@peterblazejewicz thanks for comments. Few changes suggested by you broke the tests. Can you advice please?

```
DefinitelyTyped/types/abstract-things/thing.d.ts:21:56
ERROR: 21:56  expect  TypeScript@4.6 compile error: 
Cannot find name 'EmitEventOptions'.
```

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@Shaquu thx!

File: types/abstract-things/thing.d.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    get id(): string | null;
+    set id(identifier: string | null);
+
+    init(): Promise<undefined | this>;
+    initCallback(): Promise<void>;
+
+    destroy(): Promise<void>;
+    destroyCallback(): Promise<void>;
+
+    emitEvent(event: string, data?: unknown, options?: EmitEventOptions): void;
```
Comment (andrewbranch) [inline types/abstract-things/thing.d.ts:21]:
```suggestion
    emitEvent(event: string, data?: unknown, options?: Thing.EmitEventOptions): void;
```

**Review by andrewbranch — CHANGES_REQUESTED** [review]
I fixed your code compilation error, but you still have some test failures to fix.

**Comment by Shaquu** [comment]
@andrewbranch I applied a fix for tests. Thanks for your input.

**Review by andrewbranch — APPROVED** [review]

**Comment by Shaquu** [comment]
Ready to merge

**Pull request merged**